File: algorithms/graph_algorithms/bfs/bfs.py
# Breadth-First Search: Level-by-level exploration
from collections import deque


def bfs_iterative(graph, start, goal):
    """Standard BFS with queue - finds shortest path."""
    if start == goal:
        return [start]
    
    queue = deque([(start, [start])])
    visited = {start}
    
    while queue:
        node, path = queue.popleft()
        for neighbor in graph.get(node, []):
            if neighbor in visited:
                continue
            if neighbor == goal:
                return path + [neighbor]
            visited.add(neighbor)
            queue.append((neighbor, path + [neighbor]))
    
    return None


def bfs_level_order(graph, start):
    """BFS that returns nodes level by level."""
    levels = []
    queue = deque([start])
    visited = {start}
    
    while queue:
        level_size = len(queue)
        level = []
        for _ in range(level_size):
            node = queue.popleft()
            level.append(node)
            for neighbor in graph.get(node, []):
                if neighbor not in visited:
                    visited.add(neighbor)
                    queue.append(neighbor)
        levels.append(level)
    
    return levels


def bfs_bidirectional(graph, start, goal):
    """BFS from both ends - requires reverse graph for directed graphs."""
    if start == goal:
        return [start]
    
    # Build reverse graph for backward search
    reverse_graph = {}
    for node in graph:
        for neighbor in graph[node]:
            if neighbor not in reverse_graph:
                reverse_graph[neighbor] = []
            reverse_graph[neighbor].append(node)
    
    queue_front = [(start, [start])]
    queue_back = [(goal, [goal])]
    visited_front = {start: [start]}
    visited_back = {goal: [goal]}
    
    while queue_front and queue_back:
        # Forward step
        if queue_front:
            node, path = queue_front.pop(0)
            for neighbor in graph.get(node, []):
                if neighbor in visited_back:
                    return path + list(reversed(visited_back[neighbor]))
                if neighbor not in visited_front:
                    new_path = path + [neighbor]
                    visited_front[neighbor] = new_path
                    queue_front.append((neighbor, new_path))
        
        # Backward step (using reverse graph)
        if queue_back:
            node, path = queue_back.pop(0)
            for neighbor in reverse_graph.get(node, []):
                if neighbor in visited_front:
                    return visited_front[neighbor] + list(reversed(path))
                if neighbor not in visited_back:
                    new_path = path + [neighbor]
                    visited_back[neighbor] = new_path
                    queue_back.append((neighbor, new_path))
    
    return None

File: algorithms/graph_algorithms/bfs/bfs.rs
// Breadth-First Search: Level-by-level exploration
use std::collections::{HashMap, HashSet, VecDeque};


fn bfs_iterative(graph: &HashMap<&str, Vec<&str>>, start: &str, goal: &str) -> Option<Vec<String>> {
    if start == goal {
        return Some(vec![start.to_string()]);
    }

    let mut queue = VecDeque::new();
    let mut visited = HashSet::new();
    
    queue.push_back((start, vec![start.to_string()]));
    visited.insert(start);

    while let Some((node, path)) = queue.pop_front() {
        for &neighbor in graph.get(node).unwrap_or(&vec![]) {
            if visited.contains(neighbor) {
                continue;
            }
            if neighbor == goal {
                let mut result = path.clone();
                result.push(neighbor.to_string());
                return Some(result);
            }
            visited.insert(neighbor);
            let mut new_path = path.clone();
            new_path.push(neighbor.to_string());
            queue.push_back((neighbor, new_path));
        }
    }

    None
}


fn bfs_level_order(graph: &HashMap<&str, Vec<&str>>, start: &str) -> Vec<Vec<String>> {
    let mut levels = Vec::new();
    let mut queue = VecDeque::from([start]);
    let mut visited = HashSet::from([start]);

    while !queue.is_empty() {
        let level_size = queue.len();
        let mut level = Vec::new();
        for _ in 0..level_size {
            let node = queue.pop_front().unwrap();
            level.push(node.to_string());

            for &neighbor in graph.get(node).unwrap_or(&vec![]) {
                if !visited.contains(neighbor) {
                    visited.insert(neighbor);
                    queue.push_back(neighbor);
                }
            }
        }
        levels.push(level);
    }

    levels
}


fn bfs_bidirectional(graph: &HashMap<&str, Vec<&str>>, start: &str, goal: &str) -> Option<Vec<String>> {
    // Bidirectional BFS: Search from both start and goal simultaneously
    // Note: For directed graphs, we need a reverse graph for backward search
    if start == goal {
        return Some(vec![start.to_string()]);
    }

    // Build reverse graph for backward traversal
    let mut reverse_graph: HashMap<&str, Vec<&str>> = HashMap::new();
    for (&node, neighbors) in graph {
        for &neighbor in neighbors {
            reverse_graph.entry(neighbor).or_insert_with(Vec::new).push(node);
        }
    }

    let mut queue_front = VecDeque::from([(start, vec![start.to_string()])]);
    let mut queue_back = VecDeque::from([(goal, vec![goal.to_string()])]);
    let mut visited_front: HashMap<&str, Vec<String>> = HashMap::new();
    let mut visited_back: HashMap<&str, Vec<String>> = HashMap::new();
    
    visited_front.insert(start, vec![start.to_string()]);
    visited_back.insert(goal, vec![goal.to_string()]);

    while !queue_front.is_empty() && !queue_back.is_empty() {
        // Forward step
        if let Some((node, path)) = queue_front.pop_front() {
            for &neighbor in graph.get(node).unwrap_or(&vec![]) {
                if let Some(back_path) = visited_back.get(neighbor) {
                    let mut result = path.clone();
                    result.extend(back_path.iter().rev().cloned());
                    return Some(result);
                }
                if !visited_front.contains_key(neighbor) {
                    let mut new_path = path.clone();
                    new_path.push(neighbor.to_string());
                    visited_front.insert(neighbor, new_path.clone());
                    queue_front.push_back((neighbor, new_path));
                }
            }
        }

        // Backward step (using reverse graph)
        if let Some((node, path)) = queue_back.pop_front() {
            for &neighbor in reverse_graph.get(node).unwrap_or(&vec![]) {
                if let Some(front_path) = visited_front.get(neighbor) {
                    let mut result = front_path.clone();
                    result.extend(path.iter().rev().cloned());
                    return Some(result);
                }
                if !visited_back.contains_key(neighbor) {
                    let mut new_path = path.clone();
                    new_path.push(neighbor.to_string());
                    visited_back.insert(neighbor, new_path.clone());
                    queue_back.push_back((neighbor, new_path));
                }
            }
        }
    }

    None
}


fn main() {
    let mut graph = HashMap::new();
    graph.insert("A", vec!["B", "C"]);
    graph.insert("B", vec!["D", "E"]);
    graph.insert("C", vec!["F"]);
    graph.insert("D", vec![]);
    graph.insert("E", vec!["F"]);
    graph.insert("F", vec![]);
    
    println!("{:?}", bfs_iterative(&graph, "A", "F"));
    println!("{:?}", bfs_level_order(&graph, "A"));
    println!("{:?}", bfs_bidirectional(&graph, "A", "F"));
}
